# HG changeset patch
# User John W. Eaton <jwe@octave.org>
# Date 1340744028 14400
#      Tue Jun 26 16:53:48 2012 -0400
# Node ID d40349493d874022ed9187e729c6b25e7aa53b89
# Parent  2e2a336c6b4813b1d6c64142336e7e344f11dce7
provide template version of octave_NA

* lo-ieee.h (octave_numeric_limits): New template for NA value.
Provide specializations for double and float.

diff --git a/liboctave/lo-ieee.h b/liboctave/lo-ieee.h
--- a/liboctave/lo-ieee.h
+++ b/liboctave/lo-ieee.h
@@ -113,9 +113,31 @@ extern OCTAVE_API int __lo_ieee_float_si
 
 #define lo_ieee_is_NA(x) (sizeof (x) == sizeof (float) ? \
                           __lo_ieee_float_is_NA (x) : __lo_ieee_is_NA (x))
 #define lo_ieee_is_NaN_or_NA(x) (sizeof (x) == sizeof (float) ? \
                           __lo_ieee_float_is_NaN_or_NA (x) : __lo_ieee_is_NaN_or_NA (x))
 #define lo_ieee_signbit(x) (sizeof (x) == sizeof (float) ? \
                           __lo_ieee_float_signbit (x) : __lo_ieee_signbit (x))
 
+#ifdef __cplusplus
+
+template <typename T>
+struct octave_numeric_limits
+{
+  static T NA (void) { return static_cast<T> (0); }
+};
+
+template <>
+struct octave_numeric_limits<double>
+{
+  static double NA (void) { return octave_NA; }
+};
+
+template <>
+struct octave_numeric_limits<float>
+{
+  static float NA (void) { return octave_Float_NA; }
+};
+
 #endif
+
+#endif
